csharp/src/Google.Protobuf/CodedOutputWriter.cs

```diff
@@ -0,0 +1,652 @@
+                }
+                else
+                {
+                    ReadOnlySpan<char> source = value.AsSpan();
+
+                    int bytesUsed;
+
+                    unsafe
+                    {
+                        fixed (char* sourceChars = &MemoryMarshal.GetReference(source))
+                        fixed (byte* destinationBytes = &MemoryMarshal.GetReference(buffer))
+                        {
+                            bytesUsed = Encoding.UTF8.GetBytes(sourceChars, source.Length, destinationBytes, buffer.Length);
+                        }
+                    }
+
+                    buffered += bytesUsed;
+                }
+            }
+            else
+            {
+                // The destination byte array might not be large enough so multiple writes are sometimes required
+                if (encoder == null)
+                {
+                    encoder = Encoding.UTF8.GetEncoder();
+                }
+
+                ReadOnlySpan<char> source = value.AsSpan();
+                int written = 0;
+
+                while (true)
+                {
+                    int bytesUsed;
+                    int charsUsed;
+
+                    unsafe
+                    {
+                        fixed (char* sourceChars = &MemoryMarshal.GetReference(source))
+                        fixed (byte* destinationBytes = &MemoryMarshal.GetReference(buffer))
+                        {
+                            encoder.Convert(sourceChars, source.Length, destinationBytes, buffer.Length, false, out charsUsed, out bytesUsed, out _);
+                        }
+                    }
+
+                    source = source.Slice(charsUsed);
+                    written += bytesUsed;
+
+                    buffered += bytesUsed;
+
+                    if (source.Length == 0)
+                    {
+                        break;
+                    }
+
+                    Ensure(length - written);
+                    buffer = span.Slice(buffered);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Writes a message, without a tag, to the stream.
+        /// The data is length-prefixed.
+        /// </summary>
+        /// <param name="value">The value to write</param>
+        public void WriteMessage(IBufferMessage value)
+        {
+            WriteLength(value.CalculateSize());
+            value.WriteTo(ref this);
+        }
+
+        /// <summary>
+        /// Writes a group, without a tag, to the stream.
+        /// </summary>
+        /// <param name="value">The value to write</param>
+        public void WriteGroup(IBufferMessage value)
+        {
+            value.WriteTo(ref this);
+        }
+
+        /// <summary>
+        /// Write a byte string, without a tag, to the stream.
+        /// The data is length-prefixed.
+        /// </summary>
+        /// <param name="value">The value to write</param>
+        public void WriteBytes(ByteString value)
```

Comment:
Could an overload that takes a ReadOnlySpan<byte> also be provided? That way I don't have to call ToArray() to pass a buffer to ByteString.

Reply:
It would only be called if you were using the writer directly. Generated code will always output a ByteString.

This could be added in the future but this PR has enough without more new features. You should create an issue to look at it afterwards.

Reply:
Thinking about it more, would it make more sense to replace the ByteString overload with a span, then update the generated code to pass the Span property on ByteString? That's a pretty breaking change though.

My use case is custom serialization (no code generation) only for 1 message type (for performance reasons) that involves rather large byte arrays I'd like to pool (https://github.com/apache/kudu/blob/master/src/kudu/common/wire_protocol.proto#L203)

It's a very specific use case, I'll open a separate issue later. Thanks for working on all this performance stuff!

Reply:
> Thinking about it more, would it make more sense to replace the ByteString overload with a span, then update the generated code to pass the Span property on ByteString? That's a pretty breaking change though.

WriteBytes is already copying from the the ByteString internal data to the buffer writer:

```cs
/// <summary>
/// Write a byte string, without a tag, to the stream.
/// The data is length-prefixed.
/// </summary>
/// <param name="value">The value to write</param>
public void WriteBytes(ByteString value)
{
    WriteLength(value.Length);

    Write(value.Span);
}
```

Changing WriteBytes to take a span doesn't do anything.

What you're not considering is the data passed to the writer has to come from somewhere. At the moment ByteString will always copy whatever data is passed to it, then will write that with the rest of the message.

If you want to pool then you have to consider lifetime:
1. Rent an array
2. Use it to create a ByteString (without a copy - not possible at the moment)
3. Message gets written by CodedOutputWriter
4. Array is returned to the pool (who would do this?)
5. ByteString is no longer safe to use because the underlying array is now used somewhere else.

Reply:
Thanks for the info. I haven't thought about it much, in pseudo code I was planning to do something like,
```csharp
var size = ...;
var buffer = ArrayPool<byte>.Shared.Rent(size);
try
{
    EncodeOperations(rows, buffer);
    var protoMessage = new Message()
    {
        OtherStuff = 123,
        Data = buffer.Slice(0, size);
    };
    await SendAsync(protoMessage);
}
finally
{
    ArrayPool<byte>.Shared.Return(buffer);
}
```

Though maybe to save that extra copy (to the pooled buffer), I could instead use protobuf to write every field except the large buffer, then manually write my large block of data using the IBufferWriter directly. I think since protocol buffers doesn't care about field ordering it would work.